<center><h1>Last_First_HW3</h1></center>
<br>
<br>

Name: Tufei Cai
<br>
Github Username: Tufei Tsai
<br>
USC ID: 1947878719

## 1. Time Series Classification Part 1: Feature Creation/Extraction

### (a) Download Data

Package imports

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import glob
import os
import bootstrapped.bootstrap as bs
import bootstrapped.stats_functions as bs_stats
# pip3 install bootstrapped

Get the AReM Data Set

In [2]:
bending1_ds1 = pd.read_csv('../data/bending1/dataset1.csv', skiprows = 4)
bending1_ds2 = pd.read_csv('../data/bending1/dataset2.csv',skiprows = 4)
bending1_df = pd.concat([bending1_ds1, bending1_ds2], ignore_index=True)
bending1_df.head(10)
bending1_ds2.shape

(480, 7)

In [3]:
bending2_ds1 = pd.read_csv('../data/bending2/dataset1.csv', skiprows = 4)
bending2_ds2 = pd.read_csv('../data/bending2/dataset2.csv',skiprows = 4)
bending2_df = pd.concat([bending2_ds1, bending2_ds2], ignore_index=True)
bending2_df.head(10)
bending2_ds2.shape

(480, 7)

### (b) Test and Train Data

#### Test Data

In [4]:
# datasets 1, 2, and 3 in other folders as test data and other datasets as train data.

# Folders for test data: cycling, lying, sitting, standing, walking

# cycling
cycling_path = '../data/cycling'
cycling_dfs = []

for dataset_num in range(1,4):
    file_path = f"{cycling_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    cycling_dfs.append(df)

cycling_test_df = pd.concat(cycling_dfs, ignore_index=True)

# lying
lying_path = '../data/lying'
lying_csv = glob.glob(lying_path + "/*.csv")
lying_dfs = []

for dataset_num in range(1,4):
    file_path = f"{lying_path}/dataset{dataset_num}.csv"
       
    df = pd.read_csv(file_path, skiprows=4)
    
    lying_dfs.append(df)

lying_test_df = pd.concat(lying_dfs, ignore_index=True)

# sitting
sitting_path = '../data/sitting'
sitting_dfs = []

for dataset_num in range(1,4):
    file_path = f"{sitting_path}/dataset{dataset_num}.csv"
       
    df = pd.read_csv(file_path, skiprows=4)
    
    sitting_dfs.append(df)

sitting_test_df = pd.concat(sitting_dfs, ignore_index=True)

# standing 
standing_path = '../data/standing'
standing_dfs = []

for dataset_num in range(1,4):
    file_path = f"{standing_path}/dataset{dataset_num}.csv"
       
    df = pd.read_csv(file_path, skiprows=4)
    
    standing_dfs.append(df)

standing_test_df = pd.concat(standing_dfs, ignore_index=True)

# walking
walking_path = '../data/walking'
walking_dfs = []

for dataset_num in range(1,4):
    file_path = f"{walking_path}/dataset{dataset_num}.csv"
       
    df = pd.read_csv(file_path, skiprows=4)
    
    walking_dfs.append(df)

walking_test_df = pd.concat(walking_dfs, ignore_index=True)

test_dfs = [cycling_test_df, lying_test_df, sitting_test_df,standing_test_df, walking_test_df]
df_test = pd.concat(test_dfs, ignore_index=True)

In [5]:
df_test.shape # 15 * 480

(7200, 7)

#### Train Data

In [6]:
# bending1: 3-7
# beding2: 3-6
# cycling: 4-15
# lying: 4-15
# sitting: 4-15
# standing: 4-15
# walking: 4-15

# bending1
bending1_path = '../data/bending1'
bending1_train_dfs = []

for dataset_num in range(3,8):
    file_path = f"{bending1_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    bending1_train_dfs.append(df)

bending1_train_df = pd.concat(bending1_train_dfs, ignore_index=True)

# bending2
bending2_path = '../data/bending2'
bending2_train_dfs = []

for dataset_num in range(3,7):
    file_path = f"{bending2_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    bending2_train_dfs.append(df)

bending2_train_df = pd.concat(bending2_train_dfs, ignore_index=True)

# lying
lying_train_dfs = []

for dataset_num in range(4,16):
    file_path = f"{lying_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    lying_train_dfs.append(df)

lying_train_df = pd.concat(lying_train_dfs, ignore_index=True)

# cycling
cycling_train_dfs = []

for dataset_num in range(4,16):
    file_path = f"{cycling_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    cycling_train_dfs.append(df)

cycling_train_df = pd.concat(cycling_train_dfs, ignore_index=True)

# sitting
sitting_train_dfs = []

for dataset_num in range(4,16):
    file_path = f"{sitting_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    sitting_train_dfs.append(df)

sitting_train_df = pd.concat(sitting_train_dfs, ignore_index=True)

# seems like it is only 479 rows with one NaN row
# We now add a NaN row to the df
nan_row = pd.Series([pd.NA, pd.NA, pd.NA,pd.NA,pd.NA,pd.NA,pd.NA], index=sitting_train_df.columns)
sitting_train_df = sitting_train_df.append(nan_row, ignore_index = True) 


# standing
standing_train_dfs = []

for dataset_num in range(4,16):
    file_path = f"{standing_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    standing_train_dfs.append(df)

standing_train_df = pd.concat(standing_train_dfs, ignore_index=True)

# walking
walking_train_dfs = []

for dataset_num in range(4,16):
    file_path = f"{walking_path}/dataset{dataset_num}.csv"
    
    df = pd.read_csv(file_path, skiprows=4)
    
    walking_train_dfs.append(df)

walking_train_df = pd.concat(walking_train_dfs, ignore_index=True)

train_dfs = [bending1_train_df, bending2_train_df, cycling_train_df, lying_train_df, sitting_train_df,standing_train_df, walking_train_df]
df_train = pd.concat(train_dfs, ignore_index=True)

In [7]:
df_train.shape # 69 x 480

(33120, 7)

### (c) Feature Extraction

#### i. Research

**Types of time-domain features in time seires:**

- Mean
- Median
- Standard Deviation
- Variance
- Minimum
- Maximum
- Range
- Kurtosis : A measure of whether the data are heavy-tailed or light-tailed relative to a normal distribution
- Absolute Mean: The average of the absolute values of the data points
- Interquartile Range(IQR)

#### ii. Extraction

In [17]:
# bending1
bending1_features = []

bending1_path = '../data/bending1'

for dataset_num in range(1, 8):
    file_path = f"{bending1_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    bending1_features.append(df_features)
    
bending1_extract = pd.DataFrame(bending1_features)
columns = ['Instance'] + [col for col in bending1_extract if col != 'Instance']
bending1_extract = bending1_extract[columns]


# bending2
bending2_features = []

bending2_path = '../data/bending2'

for dataset_num in range(1, 7):
    file_path = f"{bending2_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    bending2_features.append(df_features)
    
bending2_extract = pd.DataFrame(bending2_features)
columns = ['Instance'] + [col for col in bending2_extract if col != 'Instance']
bending2_extract = bending2_extract[columns]


# cycling
cycling_features = []

for dataset_num in range(1, 16):
    file_path = f"{cycling_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    cycling_features.append(df_features)
    
cycling_extract = pd.DataFrame(cycling_features)
columns = ['Instance'] + [col for col in cycling_extract if col != 'Instance']
cycling_extract = cycling_extract[columns]

# lying
lying_features = []

for dataset_num in range(1, 16):
    file_path = f"{lying_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    lying_features.append(df_features)
    
lying_extract = pd.DataFrame(lying_features)
columns = ['Instance'] + [col for col in lying_extract if col != 'Instance']
lying_extract = lying_extract[columns]


# Sitting
sitting_features = []

for dataset_num in range(1, 16):
    file_path = f"{sitting_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    sitting_features.append(df_features)
    
sitting_extract = pd.DataFrame(sitting_features)
columns = ['Instance'] + [col for col in sitting_extract if col != 'Instance']
sitting_extract = sitting_extract[columns]

# standing
standing_features = []

for dataset_num in range(1, 16):
    file_path = f"{standing_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    standing_features.append(df_features)
    
standing_extract = pd.DataFrame(standing_features)
columns = ['Instance'] + [col for col in standing_extract if col != 'Instance']
standing_extract = standing_extract[columns]

# walking
walking_features = []

for dataset_num in range(1, 16):
    file_path = f"{walking_path}/dataset{dataset_num}.csv"

    df = pd.read_csv(file_path, skiprows=4).drop('# Columns: time', axis=1)
    
    df_features = {
        'Instance': f'{dataset_num}'
    }
    for i, column in enumerate(df.columns, start=1):
        df_features[f'min_{i}'] = df[column].min()
        df_features[f'max_{i}'] = df[column].max()
        df_features[f'std_{i}'] = df[column].std()
        df_features[f'mean_{i}'] = df[column].mean()
        df_features[f'median_{i}'] = df[column].median()
        df_features[f'1st_quart_{i}'] = df[column].quantile(0.25)
        df_features[f'3rd_quart_{i}'] = df[column].quantile(0.75)
    
    walking_features.append(df_features)
    
walking_extract = pd.DataFrame(walking_features)
columns = ['Instance'] + [col for col in walking_extract if col != 'Instance']
walking_extract = walking_extract[columns]

In [21]:
extract_datasets = [bending1_extract, bending2_extract, cycling_extract, lying_extract, sitting_extract, standing_extract, walking_extract]

combined_df = pd.DataFrame()

instance_counter = 1

for df in extract_datasets:
    df['Instance'] = range(instance_counter, instance_counter + len(df))
    instance_counter += len(df)
    combined_df = pd.concat([combined_df, df], ignore_index=True)


cols = ['Instance'] + [col for col in combined_df.columns if col != 'Instance']
combined_df = combined_df[cols]
combined_df.set_index('Instance', inplace=True)

In [22]:
combined_df

min_1  max_1     std_1     mean_1  median_1  1st_quart_1  \
Instance                                                             
1         37.25  45.00  1.476967  40.624792     40.50        39.25   
2         38.00  45.67  1.435550  42.812812     42.50        42.00   
3         35.00  47.40  1.558835  43.954500     44.33        43.00   
4         33.00  47.75  3.670666  42.179813     43.50        39.15   
5         33.00  45.75  2.243490  41.678063     41.75        41.33   
...         ...    ...       ...        ...       ...          ...   
84        19.50  45.33  4.650935  33.586875     34.25        30.25   
85        19.75  45.50  4.752477  34.322750     35.25        31.00   
86        19.50  46.00  4.842294  34.546229     35.25        31.25   
87        23.50  46.25  4.531720  34.873229     35.25        31.75   
88        19.25  44.00  4.796705  34.473188     35.00        31.25   

          3rd_quart_1  min_2  max_2     std_2  ...  median_5  1st_quart_5  \
Instance                                       ...                          
1             42.0000    0.0   1.30  0.322605  ...     35.00      33.0000   
2             43.6700    0.0   1.22  0.289158  ...     33.00      32.0000   
3             45.0000    0.0   1.70  0.338690  ...     36.00      35.3625   
4             45.0000    0.0   3.00  0.630860  ...     35.00      30.4575   
5             42.7500    0.0   2.83  0.405469  ...     30.00      28.4575   
...               ...    ...    ...       ...  ...       ...          ...   
84            37.0000    0.0  14.67  2.461077  ...     16.00      13.7300   
85            38.0000    0.0  13.47  2.599284  ...     15.75      13.5000   
86            37.8125    0.0  12.47  2.360448  ...     15.75      14.0000   
87            38.2500    0.0  14.82  2.435437  ...     16.00      13.7500   
88            38.0000    0.0  13.86  2.388746  ...     16.00      13.7300   

          3rd_quart_5  min_6  max_6     std_6    mean_6  median_6  \
Instance                                                            
1               36.00   0.00   1.92  0.582915  0.570583      0.43   
2               34.50   0.00   3.11  0.601010  0.571083      0.43   
3               36.50   0.00   1.79  0.513506  0.493292      0.43   
4               36.33   0.00   2.18  0.524317  0.613521      0.50   
5               31.25   0.00   1.79  0.389164  0.383292      0.43   
...               ...    ...    ...       ...       ...       ...   
84              18.25   0.00   8.32  1.640243  3.259729      3.11   
85              17.75   0.00   9.67  1.732727  3.432562      3.20   
86              17.75   0.00  10.00  1.656742  3.338125      3.08   
87              18.00   0.00   9.51  1.690960  3.424646      3.27   
88              17.75   0.43   9.00  1.699114  3.340458      3.09   

          1st_quart_6  3rd_quart_6  
Instance                            
1              0.0000       1.3000  
2              0.0000       1.3000  
3              0.0000       0.9400  
4              0.0000       1.0000  
5              0.0000       0.5000  
...               ...          ...  
84             2.0500       4.3225  
85             2.1575       4.5650  
86             2.1600       4.3350  
87             2.1700       4.5000  
88             2.1200       4.3750  

[88 rows x 42 columns]

In [23]:
combined_df.describe()

min_1      max_1      std_1     mean_1   median_1  1st_quart_1  \
count  87.000000  87.000000  87.000000  87.000000  87.000000    87.000000   
mean   29.054368  45.616437   3.009985  39.021212  38.958506    37.112011   
std     9.624011   4.207745   1.771282   5.276431   5.386624     6.127846   
min     0.000000  30.000000   0.032038  24.562958  24.250000    23.187500   
25%    21.750000  44.670000   1.488923  34.904521  35.500000    31.710000   
50%    28.750000  45.750000   3.226507  39.433792  39.500000    39.250000   
75%    36.290000  47.750000   4.573261  43.067146  43.250000    42.000000   
max    48.000000  56.250000   7.684146  48.004167  48.000000    48.000000   

       3rd_quart_1  min_2      max_2      std_2  ...   median_5  1st_quart_5  \
count    87.000000   87.0  87.000000  87.000000  ...  87.000000    87.000000   
mean     41.322816    0.0   6.350920   0.995638  ...  16.219368    13.790287   
std       5.031028    0.0   5.059656   0.885875  ...   5.844368     6.122142   
min      26.500000    0.0   0.430000   0.055106  ...   2.250000     1.000000   
25%      38.250000    0.0   1.790000   0.325706  ...  13.395000    11.280000   
50%      40.250000    0.0   4.500000   0.464541  ...  16.000000    14.000000   
75%      45.000000    0.0  11.020000   1.908381  ...  17.710000    15.250000   
max      54.000000    0.0  17.240000   2.625516  ...  36.000000    35.362500   

       3rd_quart_5      min_6      max_6      std_6     mean_6   median_6  \
count    87.000000  87.000000  87.000000  87.000000  87.000000  87.000000   
mean     18.522787   0.004943   6.462644   1.055055   1.631250   1.453966   
std       5.563553   0.046101   2.533515   0.516968   1.157222   1.089858   
min       5.542500   0.000000   1.790000   0.388372   0.383292   0.430000   
25%      16.595000   0.000000   4.500000   0.618111   0.734271   0.500000   
50%      18.250000   0.000000   6.260000   0.842087   0.933000   0.830000   
75%      20.750000   0.000000   8.780000   1.643885   3.026708   2.640000   
max      36.500000   0.430000  13.610000   1.963685   3.530500   3.340000   

       1st_quart_6  3rd_quart_6  
count    87.000000    87.000000  
mean      0.906983     2.156437  
std       0.761542     1.526871  
min       0.000000     0.500000  
25%       0.430000     1.000000  
50%       0.470000     1.220000  
75%       1.680000     4.048750  
max       2.240000     4.617500  

[8 rows x 42 columns]

#### iii. Standard Deviation

In [24]:
std_dev_df = combined_df.std()
pd.DataFrame(std_dev_df)

0
min_1        9.624011
max_1        4.207745
std_1        1.771282
mean_1       5.276431
median_1     5.386624
1st_quart_1  6.127846
3rd_quart_1  5.031028
min_2        0.000000
max_2        5.059656
std_2        0.885875
mean_2       1.577908
median_2     1.413545
1st_quart_2  0.948434
3rd_quart_2  2.131337
min_3        2.954516
max_3        4.819848
std_3        0.952011
mean_3       3.976754
median_3     4.009687
1st_quart_3  4.184148
3rd_quart_3  4.153451
min_4        0.000000
max_4        2.181809
std_4        0.458612
mean_4       1.168268
median_4     1.148591
1st_quart_4  0.844919
3rd_quart_4  1.555235
min_5        6.121205
max_5        5.773355
std_5        1.012358
mean_5       5.704025
median_5     5.844368
1st_quart_5  6.122142
3rd_quart_5  5.563553
min_6        0.046101
max_6        2.533515
std_6        0.516968
mean_6       1.157222
median_6     1.089858
1st_quart_6  0.761542
3rd_quart_6  1.526871

In [25]:
def bootstrap_std(data, n_bootstrap=1000, ci=90):
    fill_value = np.nanmean(data)
    data_filled = np.where(np.isnan(data), fill_value, data)
    
    bootstrapped_stds = []
    
    for _ in range(n_bootstrap):
        sample = np.random.choice(data_filled, size=len(data_filled),replace = True)
        std = np.std(sample, ddof = 1)
        bootstrapped_stds.append(std)
        
    lower_percentile = (100 - ci)/2
    upper_percentile = 100 - lower_percentile
    confidence_interval = np.percentile(bootstrapped_stds, [lower_percentile, upper_percentile])
    return np.mean(bootstrapped_stds), confidence_interval


bootstrap_results = {}
for feature in combined_df.columns:
    mean_std, ci = bootstrap_std(combined_df[feature].values)
    bootstrap_results[feature] = {'Mean Std': mean_std, '90% CI': ci}
    
data_for_df = [{
    'Feature': feature,
    'Mean Std Dev': stats['Mean Std'],
    'CI Lower Bound': stats['90% CI'][0],
    'CI Upper Bound': stats['90% CI'][1]
} for feature, stats in bootstrap_results.items()]


bootstrap_df = pd.DataFrame(data_for_df)

bootstrap_df.set_index('Feature', inplace=True)
bootstrap_df

Mean Std Dev  CI Lower Bound  CI Upper Bound
Feature                                                  
min_1            9.529335        8.284955       10.746875
max_1            4.116925        3.139924        5.040232
std_1            1.750673        1.568061        1.939413
mean_1           5.197973        4.616585        5.805826
median_1         5.308841        4.710870        5.904146
1st_quart_1      6.049098        5.544034        6.592900
3rd_quart_1      4.953133        4.236665        5.704436
min_2            0.000000        0.000000        0.000000
max_2            4.986930        4.568902        5.380376
std_2            0.876392        0.802121        0.938452
mean_2           1.557146        1.394333        1.697899
median_2         1.396371        1.226802        1.537535
1st_quart_2      0.933563        0.824282        1.027191
3rd_quart_2      2.112245        1.901023        2.299251
min_3            2.915274        2.739795        3.079419
max_3            4.737639        4.017677        5.375106
std_3            0.929602        0.758247        1.115456
mean_3           3.900990        3.338196        4.430841
median_3         3.938684        3.323431        4.502599
1st_quart_3      4.115666        3.601559        4.621619
3rd_quart_3      4.090141        3.480413        4.663927
min_4            0.000000        0.000000        0.000000
max_4            2.152670        1.961888        2.358427
std_4            0.452725        0.419585        0.482934
mean_4           1.152824        1.075207        1.217848
median_4         1.133565        1.054464        1.192934
1st_quart_4      0.832226        0.767846        0.888487
3rd_quart_4      1.536780        1.431461        1.622148
min_5            6.002154        4.351219        7.561477
max_5            5.644344        4.692535        6.510814
std_5            0.996726        0.807901        1.213056
mean_5           5.636568        4.410071        6.793847
median_5         5.738234        4.509766        6.821444
1st_quart_5      6.010706        4.741281        7.245344
3rd_quart_5      5.458379        4.382902        6.525157
min_6            0.035854        0.000000        0.078467
max_6            2.499987        2.258024        2.769707
std_6            0.510344        0.475389        0.541647
mean_6           1.144190        1.060976        1.210528
median_6         1.080299        0.999440        1.145158
1st_quart_6      0.750287        0.687729        0.801806
3rd_quart_6      1.508386        1.402029        1.594777

#### iv. Select Features

**Three most important time domain features**
- Mean

The mean is sensitive to outliers and provides a summary to the overall data. 


- Minimum

The minimum has high standard deviation which means it potentially be the features of the dataset


- Maximum

The max_1 feature has a bootstrap mean standard deviation of 4.116925, with a confidence interval from 3.139924 to 5.140326, which indicating noticeable variability and possibly significant peaks in the data.

## 2. ISLR 3.7.4

### (a) Linear Train

If the true relationship between X and Y is linear, then we would expect that simple linear regression model fit the data better than the cubic regression. Cubic regression model would not improve the fir because the features they learned might not be the true relationship between X and Y. Since the cubic regression is more complex thus the cubic regression would have lower training RSS than the simple linear regression which might be overfitting.

### (b) Linear Test

The test data is something that the training model has never seen, thus the cubic regression model would learn training data very well but not on the test data. In other words, the simple linear regression would have a lower test RSS than the cubic regression because the cubic regression would have overfitting features from the training dataset. And since the relationship between X and Y is linear so the simple linear regression would perform better with the test RSS

### (c) Not Linear Train

If the relationship between X and Y is not linear, then the linear regression model might not having a good prediction because it is not linear relationship. The cubic model which is more complex statistical model would learn the non-linear relationship much better, and easily catch the hidden features from the data. Thus, we will expect the cubic regression would have lower training RSS than the linear regression model

### (d) Not Linear Testing

Since the relationship is not linear, and the cubic model fits better in the train RSS. Thus, the cubic regression model should still be better than the linear regression in the test dataset. However, if the training data contains too many random noise then the cubic regression model might do worse on the test data. But we still expect the cubic regression model do the better prediction than the simple linear regression.

## 3. ISLR 3.7.3 - Extra Practice 

## 3. ISLR 3.7.5 - Extra Practice 